[WiFiLED] Fix for setting Colors + white when called directly after each other

rvt opened the pull request:
Prevent flashes during classic driver color + white updates

- Fix for integer division calculation in LEDStateDTO and changed them to using BigDecimal.
- Fix for 'flashes' when two commands are send to the driver directly after each other. Now the driver will wait for 25ms for any new color changes, if not it will finally send the correct color values in one update.

This issue has also been reported here : https://community.openhab.org/t/wifiled-wifilight-binding/16739/21 

Signed-off-by: Ries van Twisk <github@rvt.dds.nl>

@xylo This is similar what happened with the fading driver where you could see little 'flashes' of color's of white's when sending the two commands for example:
```
		driver.setColor(HSBType.GREEN);
		driver.setWhite(PercentType.HUNDRED);
```
The integer issues are only been seen when the driver load's the white's and color's and a new DTO is set. The problem is that for example when white was read as 128 you get 128/255 * 100 and that will equal to 0 instead of 50.0 I fixed it to cast to double. This is what emcandrew126.

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/ClassicWiFiLEDDriver.java:
```diff
@@ -8,13 +8,19 @@
 import org.eclipse.smarthome.core.library.types.PercentType;
 import org.eclipse.smarthome.core.library.types.StringType;
 
+import java.io.IOException;
```
Is this the order according to the eclipse esh profile?

rvt replied:
I imported the EclipseESH profile into Idea to match the order of input.

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/ClassicWiFiLEDDriver.java:
```diff
@@ -25,26 +31,37 @@
  */
 public class ClassicWiFiLEDDriver extends AbstractWiFiLEDDriver {
 
+    public static final int WAIT_UPDATE_LED_FOR_MS = 25;
```
Should it be public?

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/LEDStateDTO.java:
```diff
@@ -70,18 +75,18 @@ public static LEDStateDTO valueOf(int state, int program, int programSpeed, int
-        long hue = (long) (hsv[0] * 360);
-        int saturation = (int) (hsv[1] * 100);
-        int brightness = (int) (hsv[2] * 100);
+        BigDecimal hue = BigDecimal.valueOf(hsv[0] * 360);
```
You are not actually using big decimals for the calculations, in this way they add no value.

rvt replied:
I am trying to be consistent with the rest of the code.
Since saturation and brightness also are BigDecimals.
I realise that DecimalType does accept a double, however PercentType does not.

What I want to do is get rid of the integer calculations...

martinvw replied:
But you don't get rid of that now, first the calculation is done and then you pass it to the big decimal, you don't use big decimal for calculations only for storage.

rvt replied:
True... but DecimalType will change a double to a BigDecimal anyways? Perhaps I am missing your point....

martinvw replied:
My point is:

> Fix for integer division calculation in LEDStateDTO and changed them to using BigDecimal.

That the statement of your commit message is not correct, instead you switched to float calculations and store them in a BigDecimal, that does not seem like a step forward to me. So let's take a step back and find out what the problem is which we try to solve and how we are getting there.

I assume hue stores an integer value between 0 and 360 and the others an integer value between 0 and 100, the only potential problem I see is that it does not do rounding but always uses `flooring` But even the impact of that seems minimal to me.

rvt replied:
Martin,

the issue is that non of the values are integers. LEDStateDTO inherits HSBType and uses for hue, saturation and brightness an BigDecimal.
I agree that using float for calculation wasn't the best idea... Double's would have been better but for sure not first get an integer while the original was a BigDecimal and then do floating point math....

FoHere is an example with the original code the below will output:

> 		LEDStateDTO ledState = LEDStateDTO.valueOf(0, 0, 0, 128, 4, 200, 0, 0);
> 		System.out.println(new Color(128,4,200).toString());
> 		System.out.println(ledState.getColor().toString());
> 

> java.awt.Color[r=128,g=4,b=200]
> java.awt.Color[r=124,g=4,b=199]

Perhaps the value differences are minimal, but when you start fading you will see little jumps because of these rounding errors. This get's very noticeable because LEDStateDTO will get re-loaded from the device before each call and for each 'brightness' change, you get different RGB values..

An alternative for this function could be:

>     public Color getColor() {
>         return new Color(super.getRGB());
>     }

Note:super.getRGB() will return a int representing a RGB value that newColor is more than happy to accept.

A other alternative is to remove the getColor method altogether and get the RGB using bit shifting. Usually I won't change a public API But if you are ok with it so I will be.... As far as I can see in the build is that LEDStateDTO is not exported in the OSGi bundle...

Before:
```
            byte r = (byte) (ledState.getColor().getRed() & 0xFF);
            byte g = (byte) (ledState.getColor().getGreen() & 0xFF);
            byte b = (byte) (ledState.getColor().getBlue() & 0xFF);
```

after:
```
            byte r = (byte) (ledState.getRGB()>>16 & 0xFF);
            byte g = (byte) (ledState.getRGB()>>8 & 0xFF);
            byte b = (byte) (ledState.getRGB() & 0xFF);
```
For Reference this is the complete function:
https://github.com/rvt/openhab2-addons/blob/master/addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/ClassicWiFiLEDDriver.java#L140

Hope I made myself more clear...

rvt replied:
Martin,

your assumption that hue is an integer is incorrect.
For example with the RGB values 128,4,200 the hue, saturation and brightness would be 
```
0 = 0.77210885
1 = 0.98
2 = 0.78431374
```

Within the constructor it get's multiplied by 360 but the getColor() method  later on in this class makes a division by 360 but based on integer. So you get:

0.77210885 8 360 = 277.959186

Then casting to int and dividing by 360 you get : 0.769444
Even if you properly round up or down, you don't get the exact hue value back and thus your RGB value might be off by one digit.

If you mean that the commit message should be updated, I can... I tried to be brief...

xylo replied:
I agree with both of you.  `RGBtoHSB(...)` and the multiplication `hsv[0] * 360` are done in float.  So it doesn't make a difference regarding the precision if `hue` is declared as `float` or `BigDecimal`.  But since `hue` is anyway converted to `BigDecimal` in line 81 (in `DecimalType`), there is also no difference in the performance.  So `float` and `BigDecimal` are equally fine and it's only a matter of code style which to use.  IMO the only sad thing is that there is no version of Color.RGBtoHSB that uses double values instead of float.  Float is pretty much deprecated today, since float and double arrays consume the size of memory in 64 bit JVMs and the float and double operations have even the same performance in 64 bit JVMs.  (At least that was the case when I made some performance tests with Oracle Java several years ago.)

martinvw replied:
If it's correct and expected to store the hue here with decimals (I personally do not think that a hue should be so accurate that it stores many decimals and that integer values of 0-255 give enough accuracy to display), please perform a double  calculation and store it as double, or any alternative where storage and calculation are consistent.

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/LEDStateDTO.java:
```diff
@@ -149,9 +154,9 @@ public LEDStateDTO withIncrementedProgramSpeed(int step) {
-        float hue = (float) this.getHue().intValue() / 360;
-        float saturation = (float) this.getSaturation().doubleValue() / 100;
-        float brightness = (float) this.getBrightness().intValue() / 100;
+        float hue = (float) this.getHue().doubleValue() / 360.0f;
```
Why adding all the f, I believe they are default for value with decimals. Better add a d to make it a double.

rvt replied:
Would the preference way to do 360d or 360.0 ?

I am planning to change it do:
```
        float hue = (float) (this.getHue().doubleValue() / 360.0);
        float saturation = (float) (this.getSaturation().doubleValue() / 100.0);
        float brightness = (float) (this.getBrightness().doubleValue() / 100.0);
```

I will add some unit tests around these calculations. The issues I found was that when I for example input this: 36,6,0 (RGB) I could get back 36,4,0 die to integer rounding.

martinvw replied:
I don't know by heart what kind of calculations are used if you divide a double by either an int, a float or a double. Maybe it's all the same, but your unit test or some extensive googling might show it.

rvt replied:
In reality the problem is really : the usage of .intValue() that will do rounding before the calculation.  I was trying to be consistent with the rest of the code within this method and to ensure that calculations are done using the correct type. Since the end result needed to be float I tried to keep everything as float.
I am more than happy to change it to double...

rvt replied:
here is a though...

I can get rid of getColor()... and use bit shifting in sendLEDData

So
```
            byte r = (byte) (ledState.getColor().getRed() & 0xFF);
            byte g = (byte) (ledState.getColor().getGreen() & 0xFF);
            byte b = (byte) (ledState.getColor().getBlue() & 0xFF);
```

becomes:
```
            byte r = (byte) (ledState.getRGB()>>16 & 0xFF);
            byte g = (byte) (ledState.getRGB()>>8 & 0xFF);
            byte b = (byte) (ledState.getRGB() & 0xFF);
```

getRGB() is already part of HSBType anyways... There is no need to create a Color object and do the calculations for the solo purpose of getting RGB values while HSBType already can do that...

martinvw replied:
That sounds good, will it allow us to get rid of the or some of the other calculations as well?

rvt replied:
_will it allow us to get rid of the or some of the other calculations as well?_
It will not git rid of the other calculations.

I tried to use : HSBType fromRGB(int r, int g, int b) { ..} but fromRGB will also cast to integers. see : https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/HSBType.java#L114 
I think this is a bug but would like to get educated of the reason why casting to int is done here.

example:
```
		HSBType test = HSBType.fromRGB(128,4,200);
		System.out.println((test.getRGB()>>16 & 0xff) + " " + (test.getRGB()>>8 & 0xff) + " " + (test.getRGB() & 0xff));
```

Will output 124 3 198 instead of 128,4,200

martinvw replied:
Old as in existing :-)

> I think this is a bug but would like to get educated of the reason why casting to int is done here.

Casting is maybe not correct, rounding should be better. But as mentioned before storing it as integer values feels like I would expect.

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/ClassicWiFiLEDDriver.java:
```diff
@@ -149,16 +170,28 @@ private void sendLEDData(LEDStateDTO ledState) throws IOException {
+                sendRaw(bytes);
+                return true;
+            } catch (IOException e) {
+                logger.warn("IOException", e);
```
Normally this error would be forwarded somewhere else, can you validate where the error did end up in the previous situation was the item state updated?

rvt replied:
Martin,

this item I was meaning to ask what the best way to handle this:

1) Since this run's in a separate thread I cannot send a response back to the handler. How is this normally handled within openhab?

2) I am not sure if it's the right way to directly update the item state, I assume you mean here to put it offline? What I see here locally that I often get  failed transmissions, This can be due to wifi or that they are just crappy devices... Is it ok to set a item offline and back online quickly after and how should retry be handled?

martinvw replied:
Can you take a look at the current situation, I don't know this piece of code and have no IDE at hand.

martinvw commented:
Okay maybe I need some more background, because storing the hue as an integer between 0 and 360 (the old way) sounds okay to me. Maybe you can give some examples with some reference where which problematic value is stored and causes troubles.

xylo commented:
@rvt 
Thanks for the contribution.  I really support that you try to improve the precision of the arithmetic operations by eliminating the unnecessary conversions to int and long.  I have to admit that I never really cared about the classical driver, because I never used it.  Instead I focused my work only on my own fading driver. :/

martinvw commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/LEDStateDTO.java:
```diff
@@ -70,18 +75,18 @@ public static LEDStateDTO valueOf(int state, int program, int programSpeed, int
         PercentType b = new PercentType(brightness);
-        PercentType w = new PercentType(white / 255 * 100);
-        PercentType w2 = new PercentType(white2 / 255 * 100);
+        PercentType w = new PercentType(BigDecimal.valueOf(Math.round(((double) white / 255) * 100)));
```
Flor clarity I would prefer to leave away the `BigDecimal.valueOf`, because you actually pass the int value and use a double calculation, passing the int will and makes the intent more clear. That percentType uses a `BigDecimal` as internal storage is imho an implementation details and no reason to use it here.

rvt commented:
Martin,

thank you for all your comment's and questions.

- All calculations are now done using BigDecimal(..) instead of floating point calculations.
- For clarity I changed LEDStateDTO to use composition over inheritance
- Added unit tests, please verify this. as the shellscript create_openhab_binding_test_skeleton.sh  created a broken unit test at first.
- Cleaned up LEDStateDTO a bit more to avoid white and white2 are now of PercentType instead of BigDecimal to avoid a view conversions during setting and getting.

I do strongly believe that this class shouldn't do 'magic' and as such I would really like that if I set some RGB values using a constructor I get the same values back in a setter, hence the need for it's own calculations.

martinvw commented:
Some final comments from my side:

I think that the code could profit from some constants for 100, 255 and maybe even 255/100 and 1f/100, this makes the code more readable and limits the creation of short lived objects.
The test class misses the copyright banner.
The test names do not follow naming conventions, i.e. do no start with lowercase and have no clear name intending the goal of the test, also I do not know why the stepsize is not just 1, should we not be able to Map the values in between?
You are missing some new lines at the end of files, in the GitHub changes view they show as a red mark on the end of the file, for example in the Pom.xml

@kaikreuzer can you take over from here, it takes too much of my holiday time, I have one week of holiday left, it's now really time to spent less instead of more time on openHAB 😉

rvt commented:
Martin,

please enjoy your holiday first! Kay, if you don't have time because you have other items on your plate then put me more back in the line..

I will work on Martin's suggestions and will clarify where needed.

martinvw commented:
> I will work on Martin's suggestions and will clarify where needed.

@rvt is this work on progress or ready for review? Can you update us about the status. And also do a rebase and fix the conflict, thanks!

martinvw commented:
> @rvt is this work on progress or ready for review? Can you update us about the status. And also do a rebase and fix the conflict, thanks!

Could you update us?

rvt commented:
Martin,

I have no time anymore to work on these items due to my new work hours that get's in the way from hobby.
Note, I did work on all the suggestions as far as I remember but unfortunately I cannot look into the code.
I see two options:
1) Please be patient as I will find some time later on, just not now.
2) Close this and ignore the PR if you feel they don't add enough.

Ries

martinvw commented:
@rvt patience is always an option for us our time is also limited, this is not really my job either :-)

@kaikreuzer can you check it? If you find small things I can also pick them up, @rvt since you created a PR to the openhab repo I have write access to this branch by default, do you agree on me committing something if needed.

martinvw commented:
@kaikreuzer only the rebase is missing, if you can review. I can fix any open issues.

kaikreuzer commented on addons/binding/org.openhab.binding.wifiled.test/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,24 @@
+Bundle-SymbolicName: org.openhab.binding.wifiled.test;singleton:=true
+Bundle-Vendor: openHAB
+Fragment-Host: org.openhab.binding.wifiled
+Bundle-Version: 2.2.0.qualifier
```
change to 2.3.0.qualifier

kaikreuzer commented on addons/binding/org.openhab.binding.wifiled.test/build.properties:
```diff
@@ -0,0 +1,6 @@
+output.. = target/test-classes
+bin.includes = META-INF/,\
+               .,\
+               src/,\
```
why src?

kaikreuzer commented on addons/binding/org.openhab.binding.wifiled.test/pom.xml:
```diff
@@ -0,0 +1,64 @@
+    <parent>
+        <groupId>org.openhab.binding</groupId>
+        <artifactId>pom</artifactId>
+        <version>2.2.0-SNAPSHOT</version>
```
2.3.0-SNAPSHOT

kaikreuzer commented on addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/ClassicWiFiLEDDriver.java:
```diff
@@ -137,28 +152,44 @@ public synchronized void incProgramSpeed(int step) throws IOException {
+                sendRaw(bytes);
+                return true;
+            } catch (IOException e) {
+                logger.warn("Exception occurred while sending command to LED", e);
```
Shouldn't this rather be debug and the Thing status being set to OFFLINE?

martinvw replied:
@kaikreuzer I do agree, but its a pretty big refactoring because the driver is not aware of the thingHandler, what would you prefer passing the handler of adding some listener interface?

martinvw commented:
@kaikreuzer Do you know why this would return forbidden?

https://dl.bintray.com/openhab/p2/openhab-deps-repo/1.0.19/content.xml.xz

It breaks my local and the Jenkins build.

kaikreuzer commented:
See https://community.openhab.org/t/solved-unable-to-install-using-apt-get-getting-405-not-allowed-when-updating/28296/12?u=kai

martinvw commented:
@kaikreuzer imho this is ready to be merged can you take a look please!

kaikreuzer commented:
Thanks!
